ykman/opgp.py

```diff
@@ -118,32 +201,163 @@ def reset(self):
 
     def _verify(self, pw, pin):
         try:
+            pin = pin.encode('utf-8')
             self.send_apdu(0, INS.VERIFY, 0, pw, pin)
         except APDUError:
             pw_remaining = self.get_remaining_pin_tries()[pw-PW1]
             raise ValueError('Invalid PIN, {} tries remaining.'.format(
                 pw_remaining))
 
-    def get_touch(self, key_slot):
+    @property
+    def supported_touch_policies(self):
         if self.version < (4, 2, 0):
+            return []
+        if self.version < (5, 2, 1):
+            return [TOUCH_MODE.ON, TOUCH_MODE.OFF, TOUCH_MODE.FIXED]
+        if self.version >= (5, 2, 1):
+            return [
+                TOUCH_MODE.ON, TOUCH_MODE.OFF, TOUCH_MODE.FIXED,
+                TOUCH_MODE.CACHED, TOUCH_MODE.CACHED_FIXED]
+
+    @property
+    def supports_attestation(self):
+        return self.version >= (5, 2, 1)
+
+    def get_touch(self, key_slot):
+        if not self.supported_touch_policies:
             raise ValueError('Touch policy is available on YubiKey 4 or later.')
-        data = self.send_apdu(0, INS.GET_DATA, 0, key_slot)
+        if key_slot == KEY_SLOT.ATTESTATION and not self.supports_attestation:
+            raise ValueError('Attestation key not available on this device.')
+        data = self.send_apdu(0, INS.GET_DATA, 0, key_slot.touch_position)
         return TOUCH_MODE(six.indexbytes(data, 0))
 
-    def set_touch(self, key_slot, mode, pin):
-        if self.version < (4, 2, 0):
+    def set_touch(self, key_slot, mode, admin_pin):
+        if not self.supported_touch_policies:
             raise ValueError('Touch policy is available on YubiKey 4 or later.')
-        self._verify(PW3, pin)
-        self.send_apdu(0, INS.PUT_DATA, 0, key_slot,
+        if mode not in self.supported_touch_policies:
+            raise ValueError('Touch policy not available on this device.')
```

Comment:
Can we communicate which touch policy is not working, maybe something like this?

```suggestion
            raise ValueError('Touch policy {} requested, device supports only: {}'.format(mode, self.supported_touch_policies))
```